lib/SILOptimizer/Mandatory/TFUtilities.cpp

```diff
@@ -221,27 +222,19 @@ SILType tf::convertElementTypeToTensorValueType(SILType ty) {
   if (isTensorFlowValue(ty))
     return ty;
 
-  return convertElementTypeToTensorValueType(ty.getASTType(), ty.getASTContext());
+  return convertElementTypeToTensorValueType(ty.getASTType(),
+                                             ty.getASTContext());
 }
 
 /// Looks up a function in the current module. If it exists, returns it.
 /// Otherwise, attempt to link it from imported modules. Returns null if such
 /// function name does not exist.
 SILFunction *tf::lookupOrLinkFunction(StringRef name, SILModule &module) {
-  auto *fn = module.lookUpFunction(name);
-  if (!fn)
-    fn = module.findFunction(name, SILLinkage::PublicExternal);
-  assert(fn);
-  if (fn->isExternalDeclaration()) {
-    bool loaded = module.loadFunction(fn);
-    assert(loaded);
-    (void)loaded;
-    // linkFunction() can return false if this function has already been
-    // deserialized and its body available in `module`.
-    module.linkFunction(fn);
```

Comment:
What does link do? (to confirm we don't need it here). once confirmed, consider renaming the func to maybe `lookupOrLoadFunction`.

It seems we just need to load (and not link) because loading gives us the deserialized SIL code, which we can then use for generic specialization purposes. is that correct?

Reply:
Yes. `SILModule::loadFunction` turns a declaration into a deserialized definition.

Reply:
Right that's what loadFunction does. I'm curious what link does -- from quick code reading it seems it'll load/deserialize any dependencies of the func, which we won't need for generic / conformance purposes. So it should be safe to remove the linkFunction() call here.

Reply:
Link creates a declaration in the current module.

Actually this function should not take care of loading - it's failing some partitioner tests that are calling `findSILFunctionForRequiredProtocolMember`. In those cases, just a linked declaration would work. There's no need to do extra work to link the function.

Changed to calling `loadFunction` within `getSingleSubstitutionMapForFunction` when the linked function is a declaration.

Reply:
Thanks for the context.

Why does calling "load" in this function cause problem?

Here is my assessment: `lookupOrLinkFunction()` currently has two callers: `findSILFunctionForRequiredProtocolMember()` and `getSingleSubstitutionMapForFunction()`. My code reading is that when `getSingleSubstitutionMapForFunction()` calls `lookupOrLinkFunction()` with some function `foo`, `findSILFunctionForRequiredProtocolMember()` must have called `lookupOrLinkFunction()` with the same `foo` earlier.

So one simple workflow could be:
1. let `findSILFunctionForRequiredProtocolMember()` call the "full form" of `lookupOrLinkFunction()`, which does lookup, link and load if needed. When lookupOrLinkFunction() returns, it guarantees that `fn->isDefinition()` is true for foo -- I believe that invariant is required by its callers.

We just need to assert on either `fn->isDefinition()` or `!isExternalDeclaration()`, because `isDefinition()` is implemented with:
```swift
bool isDefinition() const { return !isExternalDeclaration(); }
```

2. In `getSingleSubstitutionMapForFunction()`, it just lookups the function, and then asserts  `fn->isDefinition()` is true -- because `findSILFunctionForRequiredProtocolMember()` must have made it so.

WDYT?

Reply:
> My code reading is that when getSingleSubstitutionMapForFunction() calls lookupOrLinkFunction() with some function foo, findSILFunctionForRequiredProtocolMember() must have called lookupOrLinkFunction() with the same foo earlier.
> When lookupOrLinkFunction() returns, it guarantees that fn->isDefinition() is true for foo -- I believe that invariant is required by its callers.

This is not the case however. `convertScalarToHostTensorHandle` is calling `findFunction` directly to get the function from the `TensorFlow` module, not `lookupOrLinkFunction`. 
https://github.com/apple/swift/blob/f9093894d821cd7ae7bb2c40d64f9cc7c230f91d/lib/SILOptimizer/Mandatory/TFPartition.cpp#L3614

When **calling** a generic function, the caller does not need the definition. In all cases of calls in TFPartition, there is no need to deserialize anything.

The only thing that requires a definition is getting a substitution map. Note that this does not require the definition to be deserialized in the user module. The simplest way is: 
* Always look up original definitions by calling `originalModule.lookUpFunction`.
* Whenever it needs to be called, make sure to call `linkFunction` on it and call the declaration in the current module.

With this, the helper method can go away completely.

Reply:
I believe the callers of `lookupOrLinkFunction()` are sends/recvs functions, which always need a generic env, and thus need the func definition.

Agree that not having to deserialize the func def into the user module would be nice!

do you plan to make those improvements in this PR? Or pls feel free to file a bug for me to follow up.

Reply:
> I believe the callers of `lookupOrLinkFunction()` are sends/recvs functions, which always need a generic env, and thus need the func definition.

Yes, and there's already an existing helper function for that.

> Agree that not having to deserialize the func def into the user module would be nice! do you plan to make such improvements in this PR? Or pls feel free to file a bug for me to follow up.

It's done!